linux/library/src/flutter_embedder.cc

```diff
@@ -11,18 +11,79 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-#include "include/flutter_embedder.h"
+#include <flutter/flutter_embedder.h>
 
-#include <GL/gl.h>
+#include <X11/Xlib.h>
 #include <assert.h>
+#include <gtk/gtk.h>
+#include <json/json.h>
+
 #include <chrono>
+#include <cstdlib>
+#include <iostream>
+#include <memory>
+#include <string>
 
-#include "include/embedder.h"
+#include <flutter/embedder_extensions.h>
+#include <flutter/file_chooser_plugin.h>
+#include <flutter/plugin_handler.h>
 
 static_assert(FLUTTER_ENGINE_VERSION == 1, "");
 
+// Struct for storing state within an instance of the GLFW Window.
+struct FlutterEmbedderState {
+  FlutterEngine engine;
+  std::unique_ptr<flutter::PluginHandler> plugin_handler;
+};
+typedef FlutterEmbedderState *FlutterEmbedderStatePtr;
+
 static constexpr char kDefaultWindowTitle[] = "Flutter";
 
+static void GLFWClearEventCallbacks(GLFWwindow *window);
+static void GLFWAssignEventCallbacks(GLFWwindow *window);
+static void GLFWwindowSizeCallback(GLFWwindow *window, int width, int height);
+
+static void GLFWOnFlutterPlatformMessage(const FlutterPlatformMessage *message,
+                                         void *user_data) {
+  if (message->struct_size != sizeof(FlutterPlatformMessage)) {
+    std::cerr << "Invalid message size received. Expected: "
+              << sizeof(FlutterPlatformMessage) << " but received "
+              << message->struct_size << std::endl;
+    return;
+  }
+  GLFWwindow *window = reinterpret_cast<GLFWwindow *>(user_data);
+  Json::Reader parser;
+  Json::Value json;
+  std::string channel(message->channel);
+  std::string raw_string(reinterpret_cast<const char *>(message->message), 0,
+                         message->message_size);
+  bool parsing_successful = parser.parse(raw_string, json);
+  if (!parsing_successful) {
+    std::cerr << "Unable to parse platform message" << std::endl;
+    return;
+  }
+  auto state = reinterpret_cast<FlutterEmbedderStatePtr>(
+      glfwGetWindowUserPointer(window));
```

Comment:
This line of code is long and frequent enough that I'd suggest a GetEmbedderStateForWindow(window) helper that abstracts the details.

Reply:
Done.